x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/SecondaryAuthenticator.java

```diff
@@ -0,0 +1,129 @@
+     *                 If the secondary authentication credentials are found in the thread context, but fail to be authenticated, then
+     *                 the failure is returned through {@link ActionListener#onFailure(Exception)}.
+     */
+  @Override
+  public void authenticateAndAttachToContext(RestRequest request, ActionListener<SecondaryAuthentication> listener) {
+      final ThreadContext threadContext = securityContext.getThreadContext();
+      authenticate(authListener -> authenticationService.authenticate(request, false, authListener),
+          ActionListener.wrap(secondaryAuthentication -> {
+                  if (secondaryAuthentication != null) {
+                      secondaryAuthentication.writeToContext(threadContext);
+                  }
+                  listener.onResponse(secondaryAuthentication);
+              },
+              listener::onFailure));
+  }
+
+    private void authenticate(Consumer<ActionListener<Authentication>> authenticate, ActionListener<SecondaryAuthentication> listener) {
+        final ThreadContext threadContext = securityContext.getThreadContext();
+        final String header = threadContext.getHeader(SECONDARY_AUTH_HEADER_NAME);
+        if (Strings.isNullOrEmpty(header)) {
+            logger.trace("no secondary authentication credentials found (the [{}] header is [{}])", SECONDARY_AUTH_HEADER_NAME, header);
+            listener.onResponse(null);
+            return;
+        }
+
+        final Supplier<ThreadContext.StoredContext> originalContext = threadContext.newRestorableContext(false);
+        final ActionListener<Authentication> authenticationListener = new ContextPreservingActionListener<>(originalContext,
+            ActionListener.wrap(
+                authentication -> {
+                    if (authentication == null) {
+                        logger.debug("secondary authentication failed - authentication service returned a null authentication object");
+                        listener.onFailure(new ElasticsearchSecurityException("Failed to authenticate secondary user"));
+                    } else {
+                        logger.debug("secondary authentication succeeded [{}]", authentication);
+                        listener.onResponse(new SecondaryAuthentication(securityContext, authentication));
+                    }
+                },
+                e -> {
+                    logger.debug("secondary authentication failed - authentication service responded with failure", e);
+                    listener.onFailure(new ElasticsearchSecurityException("Failed to authenticate secondary user", e));
+                }
+            ));
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            logger.trace("found secondary authentication credentials, placing them in the internal [{}] header for authentication",
+                UsernamePasswordToken.BASIC_AUTH_HEADER);
```

Comment:
Unrelated to this change but we should move `BASIC_AUTH_HEADER` away from `UsernamePasswordToken` , it confused me for more time than I'd like to admit :/